docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,169 @@
+# ADR 048: Multi Tire Gas Price System
+
+## Changelog
+
+- Dec 1, 2021: Initial Draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR describes a mechanism to maintain a consensus level multi-tiered gas prices.
+
+## Context
+
+Currently, each validator configures it's own `minimal-gas-prices` in `app.yaml`. But setting a proper minimal gas price is critical to protect network from dos attack, and it's hard for all the validators to pick a sensible value, so we proposal to maintain the gas price in consensus level.
+
+Since tendermint 0.35 has supported mempool prioritization, we can take advantage of that to implement more sophisticated gas fee system.
+
+## Multi-Tier Price System
+
+We propose a multi-tier price system on consensus to provide maximum flexibility:
+
+- Tier 1: a constant gas price, which could only be modified occasionally through governance proposal.
+- Tier 2: a dynamic gas price which is adjusted according to previous block load.
+- Tier 3: a dynamic gas price which is adjusted according to previous block load.
+
+The gas price of higher tier should bigger than the lower tier.
+
+The transaction fees are charged with the exact gas price calculated on consensus.
+
+### Transaction Fields
+
+Add a new `tier` field in `Tx` to specify the preferred tier of service.
+
+Change the semantic of `gasPrice` to `gasPriceCap`, which is the maximum gas price can be charged, if the consensus gas price is larger than `gasPriceCap`, the transaction should be rejected.
```

Comment:
So the consensus gas price acts as a floor?

Reply:
the consensus gas price is the actual gas price user is charged, since it's dynamic, the `gasPriceCap` is to prevent that price is unexpectedly high.

Reply:
should we rephrase:
"the transaction should be rejected" -> "the transaction won't be included and will stay in the mempool until the gas will go to the `gasPriceCap` level. The mempool can eventually prune old transactions"

Reply:
yeah, that's better, changed.
implementation-wise, that'll need abci++ right?

Reply:
I don't think so, the mempool management is already there. Let's double check with @cmwaters